src/logql-query.ts

```diff
@@ -1,48 +1,250 @@
-        : [];
-    this.pipeline =
-      parts && parts[2].length > 0
-        ? parts[2].split('|').filter(notEmptyString).map(trim)
-        : [];
+    const syntaxTree = parser.parse(query);
+    let parsedSelectors: Array<Matcher> = [];
+    let parsedPipeline: Array<PipelineStage> = [];
+
+    syntaxTree.iterate({
+      enter(node) {
+        if (node.name) {
+          if (node.name === 'Selector' && parsedSelectors.length === 0) {
+            parsedSelectors = parseMatchers(syntaxTree, node, query);
+
+            return false;
+          } else if (
+            node.name === 'PipelineExpr' &&
+            parsedPipeline.length === 0
+          ) {
+            parsedPipeline = parsePipelineStages(syntaxTree, node, query);
+
+            return false;
+          }
+        }
+      },
+    });
+
+    this.selectors = parsedSelectors;
+    this.pipeline = parsedPipeline;
   }
 
-  appendSelector = (newSelectors: string | Array<string>) => {
-    if (Array.isArray(newSelectors)) {
-      this.selectors.push(...newSelectors);
-    } else {
-      this.selectors.push(newSelectors);
+  removeMatcher = (matcherToRemove: Partial<Matcher>) => {
+    this.selectors = this.selectors.filter((matcher) => {
+      let include = true;
+
+      if (matcherToRemove.label) {
+        include = include && matcher.label !== matcherToRemove.label;
+      }
+
+      return include;
+    });
+
+    return this;
+  };
+
+  addMatcher = (matchers?: Matcher | Array<Matcher>) => {
+    if (!matchers) {
+      return this;
     }
 
-    this.selectors = this.selectors.filter(notEmptyString).map(trim);
+    const matchersArray = Array.isArray(matchers) ? matchers : [matchers];
+
+    matchersArray.forEach((matcher) => {
+      const existingSelectorIndex = this.selectors.findIndex(
+        ({ label }) => matcher.label === label,
+      );
+
+      if (existingSelectorIndex !== -1) {
+        this.selectors[existingSelectorIndex] = matcher;
+      } else {
+        this.selectors.push(matcher);
+      }
+    });
+
+    this.selectors = this.selectors.filter(
+      (selector, index) =>
+        this.selectors.findIndex(({ label }) => selector.label === label) ===
+        index,
+    );
+
+    return this;
+  };
+
+  removePipelineStage = (
+    pipelineStageToRemove: Partial<PipelineStage>,
+    matchOptions?: { matchLabel?: string },
+  ) => {
+    this.pipeline = this.pipeline.filter((existingStage) => {
+      let include = true;
+
+      if (pipelineStageToRemove.operator) {
+        include =
+          include && existingStage.operator !== pipelineStageToRemove.operator;
+      }
+
+      if (pipelineStageToRemove.value) {
+        include =
+          include && existingStage.value !== pipelineStageToRemove.value;
+      }
+
+      if (matchOptions?.matchLabel) {
+        include =
+          (include &&
+            !existingStage.labelsInFilter?.every(
+              (label) => label.label === matchOptions.matchLabel,
+            )) ??
```

Comment:
`??` is used to handle `null` or `undefined`, but in this case the right side (`(include && ...)`) will be a boolean.